task-worklow-refactor/README.md

```diff
@@ -0,0 +1,94 @@
+# How to split one Celery task into multiple tasks
+
+Let's talk about splitting the `task.process_message` into one processing task
+and multiple tasks handling the specific events. The main benefit from this 
+will be the possibility to use more queues (for faster/slower tasks)
+ and to differentiate tasks easily (Sentry).
+
+
+## Workflow
+ - get webhook payload or message from message bus 
+ - add aditional info which can make determining the event type easier
+ (GH webhooks - X-GitHub-Event)
+ - send it to Celery as `task.process_message` - this task will take care about:
+    - parsing of the event into object of class Event (this class provides `project`, `package_config` properties,
+     which enables us to make the checks)
+    - private repository check
+    - getting the handlers which handle the event
+    - whitelist check
+    - possibly creating records in DB (explained below)
+    - (running `pre-check` here to filter out some events?)
+    - sending specific task to Celery
+  - handle specific tasks, which directly run the handlers
+    
+#### How to pass the information needed by handlers
+ Handlers are currently using the information from event (with `project` and `package_config` properties, specific for each event),
+  service config and job config object:
+ ```python
+class JobHandler(Handler):
+    def __init__(
+        self, config: ServiceConfig, job_config: Optional[JobConfig], event: Event
```

Comment:
To make it a bit easier, we can move `config` and `job_config` to the `event`.

Reply:
config and job_config should not be part of Event

Reply:
Agree with @sakalosj here. Packing up objects might help with writing less code, and making it more easy to remember the arguments to be passed to calls, but it makes it more difficult to read and understand code. And we spend much more time reading code, than writing it, so it's more important to optimise for the reading part.

If something is called "event", it should have job configuration or service configuration stored in it.

Reply:
I confused that with `package_config`, my bad...